microhapulator/Snakefile

```diff
@@ -0,0 +1,354 @@
+            data["MappingRateFullRefr"].append(frmapped / frmaptotal)
+            data["Typed"].append(num_typed_reads)
+            data["TypingRate"].append(typing_rate)
+            data["InterlocChiSq"].append(chisq)
+            data["HetTstat"].append(tstat)
+        df = pd.DataFrame(data)
+        df.to_csv(output.tsv, sep="\t", index=False, float_format="%.4f")
+
+
+rule fastqc:
+    input:
+        lambda wildcards: sorted([fq for fq in config["readfiles"] if wildcards.sample in fq]),
+    output:
+        r1="analysis/{sample}/fastqc/R1-fastqc.html",
+        r2="analysis/{sample}/fastqc/R2-fastqc.html",
+    threads: 2
+    run:
+        shell("fastqc --outdir analysis/{wildcards.sample}/fastqc/ --threads {threads} {input}")
+        outfiles = sorted(glob(f"analysis/{wildcards.sample}/fastqc/*.html"))
+        for end, outfile in enumerate(outfiles, 1):
+            outfile = Path(outfile)
+            linkfile = f"analysis/{wildcards.sample}/fastqc/R{end}-fastqc.html"
+            symlink(outfile.name, linkfile)
+
+
+rule merge:
+    input:
+        lambda wildcards: sorted([fq for fq in config["readfiles"] if wildcards.sample in fq]),
+    output:
+        mergedfq="analysis/{sample}/{sample}.extendedFrags.fastq",
+        log="analysis/{sample}/flash.log",
+    threads: 8
+    shell:
+        """
+        flash --min-overlap=25 --max-overlap=300 --allow-outies \
+            --threads={threads} --max-mismatch-density=0.25 \
+            --output-prefix analysis/{wildcards.sample}/{wildcards.sample} \
+            {input} \
+            2>&1 | tee {output.log}
+        """
+
+
+rule read_length_distributions:
+    input:
+        lambda wildcards: sorted([fq for fq in config["readfiles"] if wildcards.sample in fq]),
+        rules.merge.output.mergedfq,
+    output:
+        r1="analysis/{sample}/{sample}-r1-read-lengths.png",
+        r2="analysis/{sample}/{sample}-r2-read-lengths.png",
+        merged="analysis/{sample}/{sample}-merged-read-lengths.png",
+    run:
+        microhapulator.api.read_length_dist(
+            input[0],
+            output.r1,
+            title=f"{wildcards.sample} R1",
+        )
+        microhapulator.api.read_length_dist(
+            input[1],
+            output.r2,
+            title=f"{wildcards.sample} R2",
+        )
+        microhapulator.api.read_length_dist(
+            input[2],
+            output.merged,
+            xlabel="Length of Merged Read Pair (bp)",
+            title=wildcards.sample,
+        )
+
+
+rule copy_and_index_marker_data:
+    input:
+        tsv=config["mhdefn"],
+        fasta=config["mhrefr"],
+    output:
+        expand("marker-refr.fasta.{suffix}", suffix=("amb", "ann", "bwt", "pac", "sa")),
+        fasta="marker-refr.fasta",
+        tsv="marker-definitions.tsv",
+    shell:
+        """
+        cp {input.tsv} {output.tsv}
+        cp {input.fasta} {output.fasta}
+        bwa index {output.fasta}
+        """
```

Comment:
Because the marker reference files are so small, the pipeline just copies them into the working directory and indexes them as a matter of course.